docs/Custom_Operation_for_GPUs.md

```diff
@@ -0,0 +1,2323 @@
+            backend=None
+            call_jaxpr={ lambda ; bg:f32[4;a:8] bh:f16[4,512,512;a:8] bi:f16[512,512]
+                bj:f16[4,512,512;a:8]. let
+                bk:f16[4,512,512;a:8] bl:f16[512,512;a:8] _:f32[16,262144;a:8] = rms_norm_bwd[
+                  eps=1e-05
+                ] bj bg bh bi
+                bm:f16[512,512] = psum[axes=('a',) axis_index_groups=None] bl
+              in (bk, bm) }
+            donated_invars=(False, False, False, False)
+            global_axis_sizes=FrozenDict({'a': 8})
+            in_axes=(FrozenDict({'a': 0}), FrozenDict({'a': 0}), FrozenDict({}), FrozenDict({'a': 0}))
+            in_positional_semantics=(<_PositionalSemantics.GLOBAL: 1>, <_PositionalSemantics.GLOBAL: 1>)
+            name=transpose(rms_norm)
+            out_axes=(FrozenDict({'a': 0}), FrozenDict({}))
+            out_positional_semantics=_PositionalSemantics.GLOBAL
+            resource_env=ResourceEnv(Mesh(device_ids=array([0, 1, 2, 3, 4, 5, 6, 7]), axis_names=('a',)), ())
+            spmd_in_axes=None
+            spmd_out_axes=None
+          ] i j k bd
+          bn:f16[32,512,512] = reshape[dimensions=None new_sizes=(32, 512, 512)] be
+        in (bn, bf) }
+      name=<unnamed function>
+      out_positional_semantics=_PositionalSemantics.GLOBAL
+      out_shardings=(OpShardingSharding({devices=[8,1,1]0,1,2,3,4,5,6,7}), OpShardingSharding({replicated}))
+      resource_env=ResourceEnv(Mesh(device_ids=array([0, 1, 2, 3, 4, 5, 6, 7]), axis_names=('a',)), ())
+    ] a b
+  in (c, d) }
+```
+
+We see that `bm:f16[512,512] = psum[axes=('a',) axis_index_groups=None] bl` has been added after the call to `rms_norm_bwd` to reduce `grad_weight` across the devics on the axis `a`, but there is no `psum` for `grad_input`.
+
+This is controlled by `named_shape` passed to the `ShapedArray` construction in abstract evaluation and the axes given to `xmap`.
+
+The following code snippet from `_rms_norm_bwd_abstract` shows that `grad_input` has the exact same shape, type, and named shape as `x` does, which means `grad_input` is sharded the same way as `x`, hence no need for a `psum` for `grad_input`.
+In contrast, `grad_weight` has the same shape and type as `weight` does, but, when `weight.named_shape` is empty, `x.named_shape` is used for `grad_weight`.  In `in_axes` of our `xmap` call, `weight` has no named axis and `weight.named_shape` is empty, but `grad_weight` now has a named axis `"a"` in `grad_weight.named_shape`.
+This makes `jax.grad` insert `psum` on the axis `"a"` for `grad_weight`.
+
+```
+weight_named_shape = (
+    weight_named_shape if weight.named_shape else x.named_shape
+)
+...
+return (
+    ShapedArray(
+        x.shape, x_dtype, named_shape=x.named_shape
+    ),  # grad input
+    ShapedArray(
+        weight.shape, w_dtype, named_shape=weight_named_shape
+    ),  # grad weight
+    ....
+)
+```
+
+# Let's put it together
+
+Here is the complete code that is functional.
```

Comment:
perhaps s/that is functional// for conciseness

Reply:
Done